the-collab-lab/tcl-34-smart-shopping-list
# Il mm joining existing list

_For an example of how to fill this template out, [see this Pull Request](https://github.com/the-collab-lab/tcl-3-smart-shopping-list/pull/44)._ 

## Description

From this pull request we allow the user to share their list and access shared lists through a unique token.

- We created the `JoinList` component, which represents the form on the Welcome screen where the user can enter the token of a shared list
- In Firebase we had the `products` collection, now it is called `listTokens`. In it we store the tokens of the lists and within each document we store the products created in the list (the idea is that in the future we will no longer use the old `product` collection)
- We added the function `createListToken`, responsible for creating a new list in Firebase, where its ID is the previously implemented user's token (it was called `userToken` and is now called `listToken`)
- We refactored the `addProduct` function, which now makes updates to the registered lists, inserting new products. The products have as ID the product name (`productName`) and the information stored about them remains the same as before
- We created the `getList` function, responsible for fetching the products of a list from its unique token
- We have a new custom hook called `useList`, responsible for returning the products of a list and the search states (loading or error)
- We refactored the `ListPage`, which is now responsible for showing the list when it exists, showing the products when they are registered, or informing the user that the entered token is not valid and redirecting him to try to create a list again
- We chose not to use the `useProducts` custom hook, it was replaced by `useList`. In the future, we can make the decision to remove it from the project if the whole team agrees with this

## Related Issue

Closes #5  

## Acceptance Criteria

- [x] Cuando el usuario aún no tiene un token en localStorage, en la pantalla de entrada/inicio, se muestra un sencillo formulario que permite al usuario introducir un token.
- [x] Al introducir el token y pulsar "submit" se guarda el token en localStorage, lo que les da el control conjunto de la lista.
- [x] Al enviar, mostrar un error si el token no existe.
- [x] Si reciben un mensaje de error, permítales volver a intentarlo o crear una nueva lista.

## Type of Changes



|     | Type                       |
| --- | -------------------------- |
|    | :bug: Bug fix              |
|  ✓ | :sparkles: New feature     |
|    | :hammer: Refactoring       |
|    | :100: Add tests            |
|    | :link: Update dependencies |
|    | :scroll: Docs              |

## Updates

### Before

![image](https://user-images.githubusercontent.com/32148199/138188003-10413e43-d40f-4dde-b180-de39f8a121de.png)

### After

![image](https://user-images.githubusercontent.com/32148199/138188044-f982108d-ed8b-4af5-a810-c2f68335cfec.png)

## Testing Steps / QA Criteria

1. Start application
2. Create a new list and when accessing it you receive the message that the list is empty
3. Return to Welcome screen
4. Insert some of these list tokens in the **Share Token** input (`bonus sorb morsel`, `listen opt ghana`)
5. Be redirected to the list and verify that there are registered products
6. Return to Welcome screen
7. Insert a random/non-existent token
8. Being redirected to an empty list screen where the user is allowed to return to the Welcome screen and try again to create a list

![GIF 20-10-2021 21-01-30](https://user-images.githubusercontent.com/32148199/138188549-ec8830c4-9e12-43fb-a039-02c414be6275.gif)
